engine/src/flutter/shell/platform/linux/fl_renderer.cc

```diff
@@ -564,6 +577,69 @@ gboolean fl_renderer_present_layers(FlRenderer* self,
   return TRUE;
 }
 
+struct _FlRendererPresentLayersData {
+  FlRenderer* self;
+  FlutterViewId view_id;
+  const FlutterLayer** layers;
+  size_t layers_count;
+  gboolean res;
+  gboolean finished;
+};
+
+// Scheduled from the raster thread to run on main thread and forward
+// the arguments to fl_renderer_present_layers_on_main_thread.
+static void fl_renderer_present_layers_trampoline(gpointer user_data) {
+  _FlRendererPresentLayersData* data =
+      static_cast<_FlRendererPresentLayersData*>(user_data);
+
+  fl_renderer_make_current(data->self);
+  data->res = fl_renderer_present_layers_on_main_thread(
+      data->self, data->view_id, data->layers, data->layers_count);
+  fl_renderer_clear_current(data->self);
+
+  // Unblock the raster thread.
+  FlRendererPrivate* priv = reinterpret_cast<FlRendererPrivate*>(
+      fl_renderer_get_instance_private(data->self));
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new(&priv->mutex);
+  data->finished = TRUE;
+  g_cond_signal(&priv->cond);
+}
+
+// Called by the engine on the raster thread. Will reschedule
+// the call to fl_renderer_present_layers_on_main_thread
+// and block until it completes.
+gboolean fl_renderer_present_layers(FlRenderer* self,
+                                    FlutterViewId view_id,
+                                    const FlutterLayer** layers,
+                                    size_t layers_count) {
+  fl_renderer_clear_current(self);
+
+  _FlRendererPresentLayersData data = {
+      .self = self,
+      .view_id = view_id,
+      .layers = layers,
+      .layers_count = layers_count,
+      .res = FALSE,
+      .finished = FALSE,
+  };
+  FlRendererPrivate* priv = reinterpret_cast<FlRendererPrivate*>(
+      fl_renderer_get_instance_private(self));
+  g_autoptr(FlEngine) engine = FL_ENGINE(g_weak_ref_get(&priv->engine));
+
+  FlTaskRunner* task_runner = fl_engine_get_task_runner(engine);
+  fl_task_runner_post_callback(task_runner,
+                               fl_renderer_present_layers_trampoline, &data);
+
+  // Wait for the task to complete.
+  g_autoptr(GMutexLocker) locker = g_mutex_locker_new(&priv->mutex);
+  while (!data.finished) {
```

Comment:
finished doesn't seem to do anything - the only case that the condition is signalled is always sets it to TRUE.

Reply:
Conditions can [wake up spuriously](https://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_cond_wait.html):

> Spurious wakeups from the pthread_cond_wait() or pthread_cond_timedwait() functions may occur. 

The loop and boolean is there to make sure that the condition woke when we expect it to. Otherwise this can result in some really annoying to debug bugs :)